subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java

```diff
@@ -113,43 +128,57 @@ static ZincScalaCompiler getCompiler(CacheRepository cacheRepository, HashedClas
         return new ZincScalaCompiler(scalaInstance, scalaCompiler, new AnalysisStoreProvider());
     }
 
-    private static ClassLoader getClassLoader(ClassPath classpath) {
+    private static ClassLoader getClassLoader(ClassPath classpath, ClassLoader parent) {
         try {
             List<URL> urls = new ArrayList<URL>();
             for (File file : classpath.getAsFiles()) {
-                urls.add(file.toURI().toURL());
+                // Having the bridge in the classloader breaks zinc
+                if (!file.toString().contains("scala3-sbt-bridge")) {
+                    urls.add(file.toURI().toURL());
+                }
             }
-            return new URLClassLoader(urls.toArray(new URL[0]));
+            return new URLClassLoader(urls.toArray(new URL[0]), parent);
         } catch (Exception ee) {
             throw new RuntimeException(ee);
         }
     }
 
-    private static ClassLoader getCachedClassLoader(HashedClasspath classpath) {
+    private static boolean isScala3(String version) {
+        return version.startsWith("3.");
+    }
+
+    private static ClassLoader getCachedClassLoader(HashedClasspath classpath, ClassLoader parent) {
         try {
             return CLASSLOADER_CACHE.get(classpath.getHash(), new Callable<ClassLoader>() {
                 @Override
                 public ClassLoader call() throws Exception {
-                    return getClassLoader(classpath.getClasspath());
+                    return getClassLoader(classpath.getClasspath(), parent);
                 }
             });
         } catch (Exception ee) {
             throw new RuntimeException(ee);
         }
     }
 
-    private static ScalaInstance getScalaInstance(HashedClasspath hashedScalaClasspath) {
-        ClassLoader scalaClassLoader = getCachedClassLoader(hashedScalaClasspath);
-        String scalaVersion = getScalaVersion(scalaClassLoader);
+    private static ScalaInstance getScalaInstance(HashedClasspath hashedScalaClasspath) throws MalformedURLException {
         ClassPath scalaClasspath = hashedScalaClasspath.getClasspath();
-
         File libraryJar = findFile(ArtifactInfo.ScalaLibraryID, scalaClasspath);
-        File compilerJar = findFile(ArtifactInfo.ScalaCompilerID, scalaClasspath);
+        URL[] libraryUrls = new URL[]{libraryJar.toURI().toURL()};
```

Comment:
This should contain both scala-library and scala3-library.

Reply:
Didn't seem to break anything, but sure I can add it!

Reply:
Fixed!